kernel/src/syscall/prlimit64.rs

```diff
@@ -42,18 +42,18 @@ pub fn sys_prlimit64(
         "pid = {}, resource = {:?}, new_rlim_addr = 0x{:x}, old_rlim_addr = 0x{:x}",
         pid, resource, new_rlim_addr, old_rlim_addr
     );
-    let mut resource_limits = ctx.process.resource_limits().lock();
+    let resource_limits = ctx.process.resource_limits();
     if old_rlim_addr != 0 {
         let rlimit = resource_limits.get_rlimit(resource);
-        ctx.user_space().write_val(old_rlim_addr, rlimit)?;
+        let rlimit_raw = rlimit.get_cur_and_max();
+        ctx.user_space().write_val(old_rlim_addr, &rlimit_raw)?;
     }
     if new_rlim_addr != 0 {
-        let new_rlimit: RLimit64 = ctx.user_space().read_val(new_rlim_addr)?;
-        debug!("new_rlimit = {:?}", new_rlimit);
-        if !new_rlimit.is_valid() {
-            return_errno_with_message!(Errno::EINVAL, "invalid rlimit");
-        }
-        *resource_limits.get_rlimit_mut(resource) = new_rlimit;
+        let new_rlimit_raw: [u64; 2] = ctx.user_space().read_val(new_rlim_addr)?;
+        debug!("new_rlimit = {:?}", new_rlimit_raw);
+        resource_limits
+            .get_rlimit(resource)
+            .set_cur_and_max(new_rlimit_raw[0], new_rlimit_raw[1])?;
```

Comment:
Define a `RawRlimit64` type explicitly.

```rust
#[repr(C)]
pub struct RawRlimit64 {
    pub current: u64,
    pub max: u64
}
```

While `RawRlimit64` has the same memory layout as `[u64; 2]`, it is more readable and less error-prone to use `RawRlimit64`. With `RawRlimit64`, we embed the knowledge of the first number being the current value and the second one being the maximum value into the type, rather than spreading this knowledge all over the codebase. And the resulting user code is more readable.

```rust
        resource_limits
            .get_rlimit(resource)
            .set_cur_and_max(new_raw.current, new_raw.max)?;
```